src/Build/BackEnd/BuildManager/BuildManager.cs

```diff
@@ -974,6 +974,18 @@ public void EndBuild()
                             _overallBuildSuccess = false;
                         }
 
+                        if (loggingService is LoggingService logService && !BuildEnvironmentState.s_runningTests)
+                        {
+                            if (_overallBuildSuccess)
+                            {
+                                Debug.Assert(logService.HasLoggedErrors() != _overallBuildSuccess, "Build succeeded but logged errors.");
+                            }
+                            else
+                            {
+                                Debug.Assert(logService.HasLoggedErrors() != _overallBuildSuccess, "Build failed but did not log errors.");
+                            }
```

Comment:
I think it'd be clearer to expand this out:

```suggestion
                            if (_overallBuildSuccess)
                            {
                                Debug.Assert(logService.HasLoggedErrors() == false, "Build succeeded but logged errors.");
                            }
                            else
                            {
                                Debug.Assert(logService.HasLoggedErrors() == true, "Build failed but did not log errors.");
                            }
```